lib/pbench/agent/tool_meister.py

```diff
@@ -1518,6 +1533,39 @@ def end_tools(self, data):
                     name,
                 )
                 failures += 1
+
+        # Remove persistent tool temporary working directory on remote host
+        directory = data["directory"]
+        tool_dir = self.directories[directory]
+
+        if self._hostname == self._controller:
+            directory_to_delete = tool_dir
+        else:
+            directory_to_delete = tool_dir.parent
+
+        self.logger.debug(
+            "%s: deleting persistent tool tmp directory %s %s",
+            self._hostname,
+            self._group,
+            directory_to_delete,
+        )
+        unexpected_files = [x for x in directory_to_delete.iterdir() if x.is_file()]
+        if not unexpected_files:
+            try:
+                shutil.rmtree(directory_to_delete)
```

Comment:
It seems like we should be complaining if there are any _files_ at **any** level of the hierarchy.  That is, we can afford to ignore subdirectories if they are empty or if they contain no subdirectories which contain any files, but we need to do a recursive file tree walk ([e.g.](https://docs.python.org/3/library/os.html#os.walk)) to find that out.

And, I wonder if, after issuing a warning, we should to go ahead and delete the files anyway.  If the files actually were valuable, then I think requiring the user to make another run is probably a lesser evil than leaving behind a bunch of useless litter (in the cases where the files are _not_ valuable).

Reply:
> And, I wonder if, after issuing a warning, we should to go ahead and delete the files anyway. If the files actually were valuable, then I think requiring the user to make another run is probably a lesser evil than leaving behind a bunch of useless litter (in the cases where the files are _not_ valuable).

That's a fair point, these tmp directory is created by us and only we should have control over this directory and if we are expected to delete this directory we should delete all the files in it as well (even though we dont expect any files in there)